### ISSUE TITLE
Is it even possible to make `tagTree` lazy?

### ISSUE THREAD

**Issue by TheKing42** [ndmitchell/tagsoup#46]
https://github.com/ndmitchell/tagsoup/blob/master/Text/HTML/TagSoup/Tree.hs#L29 contains a comment saying:

```
-- | Convert a list of tags into a tree. This version is not lazy at
--   all, that is saved for version 2.
```

The thing is though, is it even possible to make this lazy? I think the answer is no for the current API. Consider this:

`(TagOpen "html" []):undefined`

Would this output

`(TagBranch "html" [] undefined):undefined`

or

`(TagLeaf $ TagOpen "html" []):undefined`

if made lazy? The only thing we know is that it is of the form

`_:_`

so that is all a lazy version could output at the moment.

There are three solutions I can think of:

1. Tell `tagTree` which tags should have closing tags, and which shouldn't, and have it raise an error or do some other behavior when this assumption is violated. No user facing types would need to be changed. This is the least safe option though.

2. Somehow merge `TagBranch` and `TagLeaf . TagOpen` (and think this was done in the past). You could have a lazy boolean flag (evaluating this boolean flag would cause the list to evaluated.) The amount of laziness possible, while still keeping it a tree, is still limited though. You could not lazily see what comes after that part of the tree. You *would* be able to scan its inner-html lazily though. This requires changing user facing types. It also sacrifices the tree like nature of the datatype somewhat.

3. Not making it lazy.

**Comment by ndmitchell** [comment]
Yeah, the current API is probably not lazy by definition, and I think 2 is probably the right solution - but I'm not super sure. I've never used this API in anger, so it's really something best directed by real use cases. Having it not be lazy at all is probably a bad solution - it effectively caps the size of documents that can be processed, and it doesn't look like it should be that hard to make it lazy.

**Comment by TheKing42** [comment]
Probably the cleanest solution to number 2 would look sort of like this:

```
data Trees = Branch {tag :: String, afterOpen :: Trees, afterClose :: Maybe Trees} | NoTrees
```

(that type is simplified of course). `afterClose` is `Nothing` if there is no closing tag. `tag` and `afterOpen` can be determined lazily. Evaluating `afterClose` to `Nothing` or `Just _` would requires tracing the document until you reach the closing tag or the end of the enclosing tag or the end of the document.

We could implement option 3 in addition to option 1 right now without changing the API. It would be a function like:

```
tagTreeLazy :: Eq str => (str -> Bool) -> [Tag str] -> [TagTree str]
```

Where the `str -> Bool` is whether or not the tag should have a closing tag. The only thing left to consider for that solution is what happens when it's wrong.

**Comment by ndmitchell** [comment]
I'm not adverse to an API change if that's what we think is needed. I'm not sure I have enough a good model for an API. Perhaps the existing data type with:

    tagTree :: Eq str => (str -> Bool) -> [Tag str] -> ([TagTree str], [str])

That gives you the pipe of `TagTree` values, plus the second component being the list of things that were expected to get shut but didn't. You would never emit a `TagClose` unless it was unmatched. Then the "sanity" condition would be that the first list doesn't contain TagClose and the second list is empty.
